File: capri/relationship_tensor.py
"""
EEF Layer 3 — Relationship Engine.

Computes pairwise and composite ecological relationships (correlations,
mutual information, ratios, and composite indices) between variables
across the spatial grid.
"""

import numpy as np
import scipy.ndimage as ndimage
from dataclasses import dataclass
from typing import List, Dict, Optional, Tuple

@dataclass
class RelationshipDescriptorMatrix:
    """Container for the output of the relationship structure extractor."""
    data: np.ndarray                          # Shape (H, W, R_features)
    feature_names: List[str]                  # Listing of what each channel represents
    variables: List[str]

class RelationshipStructureExtractor:
    """
    EEF Layer 3: Relationship Structure Extraction.

    Computes pixel-wise interactions and ecological relationships between variables:
    1. Pearson correlation (all V*(V-1)/2 pairs)
    2. Mutual Information (all V*(V-1)/2 pairs, discretized)
    3. Ecologically grounded ratios (CHL/TSM, aphy/CHL, ADG/bbp, KD490/PAR, CHL/aphy)
    4. Composite ecological indices (Trophic State, Turbidity/Eutrophication, Light Limitation, Organic Carbon Proxy)
    """

    def __init__(
        self,
        mode: str = "all",
        window_size: int = 3,
        variables: Optional[List[str]] = None,
    ):
        """
        Args:
            mode: Method to compute relationships:
                  - 'correlation': Local Pearson correlation (21 channels)
                  - 'covariance': Local covariance (21 channels)
                  - 'mutual_information': Discretized local MI (21 channels)
                  - 'ratios': 5 specific ecological ratios (5 channels)
                  - 'indices': 4 specific composite indices (4 channels)
                  - 'all': All of the above (51 channels)
            window_size: Neighborhood window size for local operations.
            variables: List of variable names matching the cube channels.
        """
        self.mode = mode
        self.window_size = window_size
        self.variables = variables or ["CHL", "aphy", "ADG", "bbp", "TSM", "PAR", "KD490"]

    def compute_local_mi(self, c1: np.ndarray, c2: np.ndarray, bins: int = 3) -> np.ndarray:
        """Computes discretized local Mutual Information using sliding window."""
        d1 = np.clip(np.floor(c1 * (bins - 0.01)), 0, bins - 1).astype(int)
        d2 = np.clip(np.floor(c2 * (bins - 0.01)), 0, bins - 1).astype(int)
        
        H, W = c1.shape
        r = self.window_size // 2
        mi_grid = np.zeros_like(c1)
        
        for i in range(H):
            for j in range(W):
                i_min, i_max = max(0, i - r), min(H, i + r + 1)
                j_min, j_max = max(0, j - r), min(W, j + r + 1)
                p1 = d1[i_min:i_max, j_min:j_max].flatten()
                p2 = d2[i_min:i_max, j_min:j_max].flatten()
                
                n = p1.size
                joint_counts = np.zeros((bins, bins))
                for x, y in zip(p1, p2):
                    joint_counts[x, y] += 1
                
                p_xy = joint_counts / n
                p_x = p_xy.sum(axis=1)
                p_y = p_xy.sum(axis=0)
                
                mi = 0.0
                for x in range(bins):
                    for y in range(bins):
                        if p_xy[x, y] > 0:
                            mi += p_xy[x, y] * np.log2(p_xy[x, y] / (p_x[x] * p_y[y] + 1e-12) + 1e-12)
                mi_grid[i, j] = mi
        
        max_mi = np.log2(bins)
        return np.clip(mi_grid / max_mi, 0.0, 1.0)

    def _rank_transform(self, channel: np.ndarray) -> np.ndarray:
        """Rank-transforms a 2D channel to [0, 1] based on value sorting."""
        flat = channel.flatten()
        temp = flat.argsort()
        ranks = np.empty_like(temp, dtype=np.float32)
        ranks[temp] = np.arange(len(flat), dtype=np.float32)
        ranks = ranks / (len(flat) - 1.0 + 1e-12)
        return ranks.reshape(channel.shape)

    def _compute_local_correlation(self, c1: np.ndarray, c2: np.ndarray, epsilon: float = 1e-5) -> np.ndarray:
        mean_1 = ndimage.uniform_filter(c1, size=self.window_size, mode='reflect')
        mean_2 = ndimage.uniform_filter(c2, size=self.window_size, mode='reflect')
        
        prod = c1 * c2
        mean_prod = ndimage.uniform_filter(prod, size=self.window_size, mode='reflect')
        local_cov = mean_prod - mean_1 * mean_2
        
        var_1 = ndimage.uniform_filter(c1**2, size=self.window_size, mode='reflect') - mean_1**2
        var_2 = ndimage.uniform_filter(c2**2, size=self.window_size, mode='reflect') - mean_2**2
        std_12 = np.sqrt(np.clip(var_1 * var_2, 0, None))
        
        local_corr = np.zeros(c1.shape, dtype=np.float32)
        valid_mask = std_12 > epsilon
        local_corr[valid_mask] = local_cov[valid_mask] / std_12[valid_mask]
        
        return (local_corr + 1.0) * 0.5  # Map to [0, 1]

    def _compute_local_covariance(self, c1: np.ndarray, c2: np.ndarray) -> np.ndarray:
        mean_1 = ndimage.uniform_filter(c1, size=self.window_size, mode='reflect')
        mean_2 = ndimage.uniform_filter(c2, size=self.window_size, mode='reflect')
        
        prod = c1 * c2
        mean_prod = ndimage.uniform_filter(prod, size=self.window_size, mode='reflect')
        local_cov = mean_prod - mean_1 * mean_2
        
        cov_min, cov_max = local_cov.min(), local_cov.max()
        if cov_max > cov_min:
            norm_cov = (local_cov - cov_min) / (cov_max - cov_min)
        else:
            norm_cov = np.zeros_like(local_cov)
        return norm_cov

    def compute_tensor(self, cube: np.ndarray) -> RelationshipDescriptorMatrix:
        """
        Computes the relationship tensor R based on the selected mode.
        
        Args:
            cube: Input observation cube of shape (H, W, V).
            
        Returns:
            RelationshipDescriptorMatrix holding the computed tensor.
        """
        H, W, V = cube.shape
        
        data_list = []
        feature_names = []

        # ── 1. Pearson Correlation ───────────────────────────────────────────
        if self.mode in ("correlation", "all"):
            for i in range(V):
                for j in range(i + 1, V):
                    name_i, name_j = self.variables[i], self.variables[j]
                    norm_corr = self._compute_local_correlation(cube[:, :, i], cube[:, :, j])
                    data_list.append(norm_corr)
                    feature_names.append(f"corr_{name_i}_{name_j}")

        # ── 2. Spearman Rank Correlation ─────────────────────────────────────
        if self.mode in ("spearman", "all"):
            # Rank-transform channels first
            ranked_channels = [self._rank_transform(cube[:, :, v]) for v in range(V)]
            for i in range(V):
                for j in range(i + 1, V):
                    name_i, name_j = self.variables[i], self.variables[j]
                    norm_spearman = self._compute_local_correlation(ranked_channels[i], ranked_channels[j])
                    data_list.append(norm_spearman)
                    feature_names.append(f"spearman_{name_i}_{name_j}")

        # ── 3. Mutual Information ─────────────────────────────────────────────
        if self.mode in ("mutual_information", "all"):
            for i in range(V):
                for j in range(i + 1, V):
                    name_i, name_j = self.variables[i], self.variables[j]
                    mi = self.compute_local_mi(cube[:, :, i], cube[:, :, j])
                    data_list.append(mi)
                    feature_names.append(f"mi_{name_i}_{name_j}")

        # ── 4. Covariance ─────────────────────────────────────────────────────
        if self.mode in ("covariance", "all"):
            for i in range(V):
                for j in range(i + 1, V):
                    name_i, name_j = self.variables[i], self.variables[j]
                    norm_cov = self._compute_local_covariance(cube[:, :, i], cube[:, :, j])
                    data_list.append(norm_cov)
                    feature_names.append(f"cov_{name_i}_{name_j}")

        # ── 5. Ecologically Grounded Ratios ───────────────────────────────────
        var_to_idx = {v.upper(): idx for idx, v in enumerate(self.variables)}
        epsilon = 1e-6
        if self.mode == "ratios":
            # CHL / TSM
            if "CHL" in var_to_idx and "TSM" in var_to_idx:
                val = (cube[:, :, var_to_idx["CHL"]] + epsilon) / (cube[:, :, var_to_idx["TSM"]] + epsilon)
                data_list.append(np.clip(val, 0.0, 10.0) / 10.0)
                feature_names.append("ratio_CHL_TSM")
                
            # aphy / CHL
            aphy_key = "APHY" if "APHY" in var_to_idx else "aphy"
            if aphy_key in var_to_idx and "CHL" in var_to_idx:
                val = (cube[:, :, var_to_idx[aphy_key]] + epsilon) / (cube[:, :, var_to_idx["CHL"]] + epsilon)
                data_list.append(np.clip(val, 0.0, 2.0) / 2.0)
                feature_names.append("ratio_aphy_CHL")
                
            # ADG / bbp
            bbp_key = "BBP" if "BBP" in var_to_idx else "bbp"
            if "ADG" in var_to_idx and bbp_key in var_to_idx:
                val = (cube[:, :, var_to_idx["ADG"]] + epsilon) / (cube[:, :, var_to_idx[bbp_key]] + epsilon)
                data_list.append(np.clip(val, 0.0, 50.0) / 50.0)
                feature_names.append("ratio_ADG_bbp")
                
            # KD490 / PAR
            if "KD490" in var_to_idx and "PAR" in var_to_idx:
                val = (cube[:, :, var_to_idx["KD490"]] + epsilon) / (cube[:, :, var_to_idx["PAR"]] + epsilon)
                data_list.append(np.clip(val, 0.0, 5.0) / 5.0)
                feature_names.append("ratio_KD490_PAR")
                
            # CHL / aphy
            if "CHL" in var_to_idx and aphy_key in var_to_idx:
                val = (cube[:, :, var_to_idx["CHL"]] + epsilon) / (cube[:, :, var_to_idx[aphy_key]] + epsilon)
                data_list.append(np.clip(val, 0.0, 20.0) / 20.0)
                feature_names.append("ratio_CHL_aphy")

        # ── 6. Composite Ecological Indices ───────────────────────────────────
        if self.mode == "indices":
            # Trophic State Proxy: CHL * PAR
            if "CHL" in var_to_idx and "PAR" in var_to_idx:
                val = cube[:, :, var_to_idx["CHL"]] * cube[:, :, var_to_idx["PAR"]]
                data_list.append(val)
                feature_names.append("index_trophic_state")
                
            # Turbidity/Eutrophication Index: TSM * KD490
            if "TSM" in var_to_idx and "KD490" in var_to_idx:
                val = cube[:, :, var_to_idx["TSM"]] * cube[:, :, var_to_idx["KD490"]]
                data_list.append(val)
                feature_names.append("index_turbidity")
                
            # Light Limitation Index: PAR / (KD490 + epsilon)
            if "PAR" in var_to_idx and "KD490" in var_to_idx:
                val = cube[:, :, var_to_idx["PAR"]] / (cube[:, :, var_to_idx["KD490"]] + epsilon)
                data_list.append(np.clip(val, 0.0, 10.0) / 10.0)
                feature_names.append("index_light_limitation")
                
            # Organic Carbon Proxy: bbp / (TSM + epsilon)
            bbp_key = "BBP" if "BBP" in var_to_idx else "bbp"
            if bbp_key in var_to_idx and "TSM" in var_to_idx:
                val = cube[:, :, var_to_idx[bbp_key]] / (cube[:, :, var_to_idx["TSM"]] + epsilon)
                data_list.append(np.clip(val, 0.0, 5.0) / 5.0)
                feature_names.append("index_organic_carbon")

        # Combine list of HxW matrices into HxWxC tensor
        data = np.stack(data_list, axis=-1).astype(np.float32)
        
        # Save relationship tensor to disk as relationship_tensor.npy
        np.save("relationship_tensor.npy", data)
        
        return RelationshipDescriptorMatrix(
            data=data,
            feature_names=feature_names,
            variables=self.variables
        )

File: packages/core/src/utils/relationshipTensor.ts
// EEF Layer 3 — Relationship Engine (TS port of capri/relationship_tensor.py
// `RelationshipStructureExtractor.compute_tensor(mode="all")` and the
// ratio/index definitions from ecological_encoder.py).
//
// Computes 193 relationship feature channels across the 20x20 grid:
//   - 91 local correlations (all i<j pairs of the 14 variables)
//   - 91 local mutual-information channels (discretized, 3 bins)
//   - 6 ecologically grounded ratios
//   - 5 composite ecological indices
//
// Output is a flat Float32Array of shape (gridSize, gridSize, 193), row-major
// over (row, col, channel).

import { DataCube, VariableName } from "../types";
import { flattenGrid, buildPaddedGrid, boxBlur3x3 } from "./gridOps";

const VARS: VariableName[] = ["CHL", "aphy", "ADG", "bbp", "TSM", "PAR", "KD490", "FLH", "CHL_disagreement", "OA01", "OA02", "OA03", "OA04", "OA05", "OA06", "OA07", "OA08", "OA09", "OA10", "OA11", "OA13"];
const EPSILON = 1e-5;

export interface RelationshipTensor {
  data: Float32Array;
  gridSize: number;
  featureNames: string[];
}

// ── Local correlation ────────────────────────────────────────────────────

function computeLocalCorrelation(
  gridI: Float32Array,
  gridJ: Float32Array,
  gridSize: number
): Float32Array {
  const paddedSize = gridSize + 2;
  const n = gridSize * gridSize;

  const paddedI = padFromFlat(gridI, gridSize);
  const paddedJ = padFromFlat(gridJ, gridSize);

  const paddedISq = new Float32Array(paddedI.length);
  const paddedJSq = new Float32Array(paddedJ.length);
  const paddedProd = new Float32Array(paddedI.length);
  for (let i = 0; i < paddedI.length; i++) {
    paddedISq[i] = paddedI[i] * paddedI[i];
    paddedJSq[i] = paddedJ[i] * paddedJ[i];
    paddedProd[i] = paddedI[i] * paddedJ[i];
  }

  const meanI = boxBlur3x3(paddedI, paddedSize, gridSize);
  const meanJ = boxBlur3x3(paddedJ, paddedSize, gridSize);
  const meanISq = boxBlur3x3(paddedISq, paddedSize, gridSize);
  const meanJSq = boxBlur3x3(paddedJSq, paddedSize, gridSize);
  const meanProd = boxBlur3x3(paddedProd, paddedSize, gridSize);

  const out = new Float32Array(n);
  for (let i = 0; i < n; i++) {
    const localCov = meanProd[i] - meanI[i] * meanJ[i];
    let varI = meanISq[i] - meanI[i] * meanI[i];
    let varJ = meanJSq[i] - meanJ[i] * meanJ[i];
    if (varI < 0) varI = 0;
    if (varJ < 0) varJ = 0;
    const stdIJ = Math.sqrt(varI * varJ);

    let localCorr = 0;
    if (stdIJ > EPSILON) {
      localCorr = localCov / stdIJ;
    }
    // Map from [-1, 1] to [0, 1]
    out[i] = (localCorr + 1.0) * 0.5;
  }
  return out;
}

// Pads a row-major (gridSize x gridSize) flat grid using reflect-mode halo=1.
function padFromFlat(grid: Float32Array, gridSize: number): Float32Array {
  const grid2d: number[][] = [];
  for (let r = 0; r < gridSize; r++) {
    const row: number[] = [];
    for (let c = 0; c < gridSize; c++) row.push(grid[r * gridSize + c]);
    grid2d.push(row);
  }
  return buildPaddedGrid(grid2d, 1, gridSize);
}

// ── Local mutual information ────────────────────────────────────────────

function computeLocalMI(gridI: Float32Array, gridJ: Float32Array, gridSize: number, bins = 3): Float32Array {
  const n = gridSize * gridSize;
  const d1 = new Int32Array(n);
  const d2 = new Int32Array(n);
  for (let i = 0; i < n; i++) {
    let v1 = Math.floor(gridI[i] * (bins - 0.01));
    let v2 = Math.floor(gridJ[i] * (bins - 0.01));
    if (v1 < 0) v1 = 0;
    if (v1 > bins - 1) v1 = bins - 1;
    if (v2 < 0) v2 = 0;
    if (v2 > bins - 1) v2 = bins - 1;
    d1[i] = v1;
    d2[i] = v2;
  }

  const out = new Float32Array(n);
  const r = 1;
  const maxMi = Math.log2(bins);

  for (let i = 0; i < gridSize; i++) {
    for (let j = 0; j < gridSize; j++) {
      const iMin = Math.max(0, i - r);
      const iMax = Math.min(gridSize, i + r + 1);
      const jMin = Math.max(0, j - r);
      const jMax = Math.min(gridSize, j + r + 1);

      const jointCounts = new Float64Array(bins * bins);
      let count = 0;
      for (let ii = iMin; ii < iMax; ii++) {
        for (let jj = jMin; jj < jMax; jj++) {
          const idx = ii * gridSize + jj;
          jointCounts[d1[idx] * bins + d2[idx]]++;
          count++;
        }
      }

      const pX = new Float64Array(bins);
      const pY = new Float64Array(bins);
      const pXY = new Float64Array(bins * bins);
      for (let x = 0; x < bins; x++) {
        for (let y = 0; y < bins; y++) {
          const p = jointCounts[x * bins + y] / count;
          pXY[x * bins + y] = p;
          pX[x] += p;
          pY[y] += p;
        }
      }

      let mi = 0;
      for (let x = 0; x < bins; x++) {
        for (let y = 0; y < bins; y++) {
          const p = pXY[x * bins + y];
          if (p > 0) {
            mi += p * Math.log2(p / (pX[x] * pY[y] + 1e-12) + 1e-12);
          }
        }
      }

      let v = mi / maxMi;
      if (v < 0) v = 0;
      if (v > 1) v = 1;
      out[i * gridSize + j] = v;
    }
  }
  return out;
}

// ── Ratios & composite indices ───────────────────────────────────────────

function computeRatio(num: Float32Array, den: Float32Array, clampMax: number): Float32Array {
  const out = new Float32Array(num.length);
  for (let i = 0; i < num.length; i++) {
    let val = (num[i] + EPSILON) / (den[i] + EPSILON);
    if (val < 0) val = 0;
    if (val > clampMax) val = clampMax;
    out[i] = val / clampMax;
  }
  return out;
}

function computeProduct(a: Float32Array, b: Float32Array): Float32Array {
  const out = new Float32Array(a.length);
  for (let i = 0; i < a.length; i++) out[i] = a[i] * b[i];
  return out;
}

function computeRatioNoEpsNum(num: Float32Array, den: Float32Array, clampMax: number): Float32Array {
  // For indices: val = num / (den + EPSILON), clip(0, clampMax) / clampMax
  const out = new Float32Array(num.length);
  for (let i = 0; i < num.length; i++) {
    let val = num[i] / (den[i] + EPSILON);
    if (val < 0) val = 0;
    if (val > clampMax) val = clampMax;
    out[i] = val / clampMax;
  }
  return out;
}

// index_physiological_stress: raw = FLH / (CHL + EPSILON);
// val = clip(raw * (1 - CHL_disagreement), 0, clampMax) / clampMax
function computePhysiologicalStress(
  flh: Float32Array,
  chl: Float32Array,
  chlDisagreement: Float32Array,
  clampMax: number
): Float32Array {
  const out = new Float32Array(flh.length);
  for (let i = 0; i < flh.length; i++) {
    const raw = flh[i] / (chl[i] + EPSILON);
    let val = raw * (1 - chlDisagreement[i]);
    if (val < 0) val = 0;
    if (val > clampMax) val = clampMax;
    out[i] = val / clampMax;
  }
  return out;
}

// ── Public API ────────────────────────────────────────────────────────────

export function relationshipChannelIndex(tensor: RelationshipTensor, featureName: string): number {
  return tensor.featureNames.indexOf(featureName);
}

export function getRelationshipChannelGrid(tensor: RelationshipTensor, channelIdx: number): Float32Array {
  const { data, gridSize, featureNames } = tensor;
  const totalChannels = featureNames.length;
  const out = new Float32Array(gridSize * gridSize);
  for (let i = 0; i < gridSize * gridSize; i++) {
    out[i] = data[i * totalChannels + channelIdx];
  }
  return out;
}

export function computeRelationshipTensor(cube: DataCube): RelationshipTensor {
  const { gridSize } = cube;
  const n = gridSize * gridSize;

  const flatGrids: Record<VariableName, Float32Array> = {} as Record<VariableName, Float32Array>;
  for (const v of VARS) {
    flatGrids[v] = flattenGrid(cube.channels[v], gridSize);
  }

  const featureNames: string[] = [];
  const channels: Float32Array[] = [];

  // 1. Correlations (36)
  for (let i = 0; i < VARS.length; i++) {
    for (let j = i + 1; j < VARS.length; j++) {
      const corr = computeLocalCorrelation(flatGrids[VARS[i]], flatGrids[VARS[j]], gridSize);
      channels.push(corr);
      featureNames.push(`corr_${VARS[i]}_${VARS[j]}`);
    }
  }

  // 2. Mutual information (36)
  for (let i = 0; i < VARS.length; i++) {
    for (let j = i + 1; j < VARS.length; j++) {
      const mi = computeLocalMI(flatGrids[VARS[i]], flatGrids[VARS[j]], gridSize);
      channels.push(mi);
      featureNames.push(`mi_${VARS[i]}_${VARS[j]}`);
    }
  }

  // 3. Ratios (6)
  channels.push(computeRatio(flatGrids.CHL, flatGrids.TSM, 10.0));
  featureNames.push("ratio_CHL_TSM");

  channels.push(computeRatio(flatGrids.aphy, flatGrids.CHL, 2.0));
  featureNames.push("ratio_aphy_CHL");

  channels.push(computeRatio(flatGrids.ADG, flatGrids.bbp, 50.0));
  featureNames.push("ratio_ADG_bbp");

  channels.push(computeRatio(flatGrids.KD490, flatGrids.PAR, 5.0));
  featureNames.push("ratio_KD490_PAR");

  channels.push(computeRatio(flatGrids.CHL, flatGrids.aphy, 20.0));
  featureNames.push("ratio_CHL_aphy");

  channels.push(computeRatio(flatGrids.FLH, flatGrids.CHL, 5.0));
  featureNames.push("ratio_FLH_CHL");

  // 4. Composite indices (5)
  channels.push(computeProduct(flatGrids.CHL, flatGrids.PAR));
  featureNames.push("index_trophic_state");

  channels.push(computeProduct(flatGrids.TSM, flatGrids.KD490));
  featureNames.push("index_turbidity");

  channels.push(computeRatioNoEpsNum(flatGrids.PAR, flatGrids.KD490, 10.0));
  featureNames.push("index_light_limitation");

  channels.push(computeRatioNoEpsNum(flatGrids.bbp, flatGrids.TSM, 5.0));
  featureNames.push("index_organic_carbon");

  channels.push(computePhysiologicalStress(flatGrids.FLH, flatGrids.CHL, flatGrids.CHL_disagreement, 5.0));
  featureNames.push("index_physiological_stress");

  // Interleave into row-major (gridSize, gridSize, 193)
  const totalChannels = featureNames.length; // 193
  const data = new Float32Array(n * totalChannels);
  for (let c = 0; c < totalChannels; c++) {
    const ch = channels[c];
    for (let i = 0; i < n; i++) {
      data[i * totalChannels + c] = ch[i];
    }
  }

  return { data, gridSize, featureNames };
}
